analytics: Fix cut off legends and range selector in existing charts.

laurynmm opened the pull request:
It appears that when we have `font-family: "Source Sans 3"` for the Plotly.js `layout` parameter, the font-family that appears in the created chart is `"Open Sans", verdana, arial, sans-serif`, not what we intended (you can see this current behavior on the CZO stats page). And this then impacts a clips layer that's set on legends and range selectors, which I think is using "Source Sans 3" to calculate the width needed since it's a narrower font?

Setting the font-family as "Open Sans, sans-serif" fixes the text from being clipped, but the font size of 14 does not work for the horizontal axis on the vertical bar charts with the width we've selected, so this also changes the general font-size for those charts to 12, while still using font-size 14 for the legends.

But changing the font to size 12 on the **Active users** chart caused a bug in the chart rerendering when the range selector sliders were moved. Fixed this by changing the `y` value for the range selector from `-.79` to `-.69`, but I'm not sure why that works.

Marking this as a draft PR for now because I feel like there are other potential bugs from these changes that I haven't yet identified, nor do I feel clear on why the changes I've made work. But I would like some feedback on what I've done so far, and thoughts on how to proceed.

Fixes #18912.

---

**Screenshots and screen captures:**

<details>
<summary>Updated charts with labels not cut off</summary>

![Screenshot from 2022-09-28 18-18-56](https://user-images.githubusercontent.com/63245456/192833055-9ba8cd1f-3f0c-4f5c-ade9-6554ebec5326.png)
</details>

---

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

timabbott commented:
I think if we're going to spend time on improving the actual chart rendering, we should look at modern JS graphing libraries; we picked plotly like 4 years ago, and it's pretty fiddly with a lot of magic constants required (and also just a lot of code -- `stats.js` shouldn't need to be 1k lines of code!); it's very possible that rewriting the charts using a more modern library would be a better use of time than fixing several individual rendering issues. 

I can't recall if @andersk had one he liked, but I seem to recall our having a chat.zulip.org thread on the topic.

laurynmm commented:
> I think if we're going to spend time on improving the actual chart rendering, we should look at modern JS graphing libraries; we picked plotly like 4 years ago, and it's pretty fiddly with a lot of magic constants required (and also just a lot of code -- `stats.js` shouldn't need to be 1k lines of code!); it's very possible that rewriting the charts using a more modern library would be a better use of time than fixing several individual rendering issues.
> 
> I can't recall if @andersk had one he liked, but I seem to recall our having a chat.zulip.org thread on the topic.

That all makes sense to me, though I couldn't find a CZO thread.

In any event, I've updated this PR to be something I'm comfortable with as a potential clean up for the broken behavior we currently have.

I've manually tested on Linux in Firefox and Chromium with the Analytics dev realm data. I feel like it would be useful for a reviewer to test on Mac in the three browsers (Firefox, Chrome and Safari).

---

**Updated screenshots (Firefox, Linux)**:

#### Messages read over time
<details>
<summary>Pre-changes</summary>

![Screenshot from 2022-09-29 16-28-51](https://user-images.githubusercontent.com/63245456/193059583-ade340d9-d5b7-449b-8d36-1cd23f6926c3.png)
</details>

<details>
<summary>Post-changes</summary>

![Screenshot from 2022-09-29 16-32-13](https://user-images.githubusercontent.com/63245456/193060003-3a4d8f36-ddfb-43e2-b4a1-88f32980938c.png)
</details>

#### Messages sent by recipient type
<details>
<summary>Pre-changes</summary>

![Screenshot from 2022-09-29 16-29-02](https://user-images.githubusercontent.com/63245456/193059524-13626d09-d065-4c2f-897b-92d07be4e8c7.png)
</details>

<details>
<summary>Post-changes</summary>

![Screenshot from 2022-09-29 16-32-03](https://user-images.githubusercontent.com/63245456/193060045-179c6eab-f36f-4802-b8bb-95d44c8ebdbd.png)
</details>

#### Active users
<details>
<summary>Pre-changes</summary>

![Screenshot from 2022-09-29 16-29-10](https://user-images.githubusercontent.com/63245456/193059471-8e89e479-235c-476b-8898-69781b741b17.png)
</details>

<details>
<summary>Post-changes</summary>

![Screenshot from 2022-09-29 16-31-45](https://user-images.githubusercontent.com/63245456/193060086-53f85091-3c7b-4987-8e88-5c443623be4c.png)
</details>

#### Messages sent by client
<details>
<summary>Pre-changes</summary>

![Screenshot from 2022-09-29 16-29-20](https://user-images.githubusercontent.com/63245456/193059423-b24616b8-ea40-49e3-8b97-2bf7a4f5380c.png)
</details>

<details>
<summary>Post-changes</summary>

![Screenshot from 2022-09-29 16-34-39](https://user-images.githubusercontent.com/63245456/193060557-981eb8bc-3cfa-4f04-82bc-354686f0de58.png)
</details>

#### Messages sent over time
<details>
<summary>Pre-changes</summary>

![Screenshot from 2022-09-29 16-29-31](https://user-images.githubusercontent.com/63245456/193059373-4f3677ec-69bf-4931-ae91-8b344b5ecfca.png)
</details>

<details>
<summary>Post-changes</summary>

![Screenshot from 2022-09-29 16-31-24](https://user-images.githubusercontent.com/63245456/193060225-142153a8-0d7a-4f82-a2c3-a0f59f11a24a.png)
</details>

timabbott commented on templates/analytics/stats.html:
```diff
@@ -67,7 +67,8 @@ <h1>{{ _("Messages sent by client") }}</h1>
                     <h1>{{ _("Active users") }}</h1>
                     <div class="button-container">
-                        <div class="buttons button-active-users">
+                        <label>{{ _("Aggregation") }}</label>
```
What's the thinking behind the addition of this label?

laurynmm replied:
Matching the other two vertical bar/line graphs (messages sent over time and messages read).

If it doesn't make sense, we could just move the buttons to the left without a label. I think it would be better to have them not sit on top of the end of the graph.

alya replied:
How about just dropping the "Aggregation" label on all the charts? I don't think it's helping me understand what's going on.

alya replied:
Moving the buttons to the left without a label works for me.

laurynmm commented:
**Updates**:
- Removed the 'Aggregation' labels from "Messages sent over time" and "Messages read over time"..
- Moved buttons to the left in "Active users" chart without label.

**Screenshots**:
*Both "Messages sent by client" and "Messages sent by recipient" are not changed from screenshots above*

<details>
<summary>Active users: pre-changes</summary>

![Screenshot from 2022-09-29 16-29-10](https://user-images.githubusercontent.com/63245456/193059471-8e89e479-235c-476b-8898-69781b741b17.png)
</details>

<details>
<summary>Active users: post-changes</summary>

![Screenshot from 2022-09-30 13-00-32](https://user-images.githubusercontent.com/63245456/193256929-62d5be97-e0fc-4a77-ad87-a0715f7b1882.png)
</details>

<details>
<summary>Messages sent over time: pre-changes</summary>

![Screenshot from 2022-09-29 16-29-31](https://user-images.githubusercontent.com/63245456/193059373-4f3677ec-69bf-4931-ae91-8b344b5ecfca.png)
</details>

<details>
<summary>Messages sent over time: post-changes</summary>

![Screenshot from 2022-09-30 13-00-46](https://user-images.githubusercontent.com/63245456/193256968-566f1e34-bae4-4fa0-9959-c0649ff0f3c3.png)
</details>

<details>
<summary>Messages read over time: pre-changes</summary>

![Screenshot from 2022-09-29 16-28-51](https://user-images.githubusercontent.com/63245456/193059583-ade340d9-d5b7-449b-8d36-1cd23f6926c3.png)
</details>

<details>
<summary>Messages read over time: post-changes</summary>

![Screenshot from 2022-09-30 13-00-57](https://user-images.githubusercontent.com/63245456/193257001-9aeca7c6-f680-4e6d-ad55-2cc9ecaf3c99.png)
</details>

laurynmm commented:
CI failing due to [puppeteer test issue](https://chat.zulip.org/#narrow/stream/43-automated-testing/topic/main.20failing/near/1439747) that's unrelated to these changes.

alya commented:
Mm, I see the problem with Russian (good catch!), but I don't think it works to move the axis settings to the top. It makes the bar charts confusing, and it would be good to be consistent across all the charts.

What if we just left-align those settings at the bottom of the chart? We can use more newlines to simplify the layout. E.g., "Me" / "Everyone" and the time-based filters can be on different lines, and filters can be on a different line from the summary stats. Can we give that a try?

alya commented:
Hm, the "Week of.." text seems to be missing from the single-column English screenshot above. Is that screenshot from an old version of the PR?

alya commented:
It looks to me like "Total messages" is on the same line as the time range selector, while "Week of.." is just above (but maybe not even a full line?). I think we should be consistent between those two placements.

alya commented:
Other than those two points, the screenshots look good to me!

laurynmm commented:
> Hm, the "Week of.." text seems to be missing from the single-column English screenshot above. Is that screenshot from an old version of the PR?

That text only appears when you hover over the chart with the mouse. When the page loads, it's not there. So I had just taken the screenshot after the page load.

> It looks to me like "Total messages" is on the same line as the time range selector, while "Week of.." is just above (but maybe not even a full line?). I think we should be consistent between those two placements.

I've updated the PR so that the alignment is better now, see screenshots below.

I've manually tested in the dev environment on Firefox and Chromium in Linux. It might be worth doing a check on MacOS with the main browsers (Firefox, Chrome and Safari). I'm pretty sure that these changes are all improvements on the current implementation, which has the two very visible issues (cutting off text in legends and squashing the chart when the button text is long in the user's selected language).

**Screenshots - Firefox/Linux**

<details>
<summary>English - double column</summary>

![Screenshot from 2022-10-07 12-50-07](https://user-images.githubusercontent.com/63245456/194538346-6a39e7ae-8df9-4426-9e14-e3c330d53ff9.png)
</details>

<details>
<summary>Russian - double column</summary>

![Screenshot from 2022-10-07 12-50-42](https://user-images.githubusercontent.com/63245456/194538325-7710f130-3597-448f-8d3b-836f43464f9b.png)
</details>

<details>
<summary>English - single column</summary>

![Screenshot from 2022-10-07 12-52-00](https://user-images.githubusercontent.com/63245456/194538298-39c985bc-a90a-44d0-857b-257042cea11e.png)
![Screenshot from 2022-10-07 12-52-11](https://user-images.githubusercontent.com/63245456/194538301-aaa2c1a7-dcce-4175-b2c9-b2a93c27c24f.png)
</details>

<details>
<summary>Russian - single column</summary>

![Screenshot from 2022-10-07 12-51-12](https://user-images.githubusercontent.com/63245456/194538264-74991d41-2ea9-4c30-a9ec-9e0fa6af22fa.png)
![Screenshot from 2022-10-07 12-51-26](https://user-images.githubusercontent.com/63245456/194538279-0c2da3e7-22f0-4fcc-b1d7-cb2264b35a3c.png)
</details>

alya commented:
> That text only appears when you hover over the chart with the mouse. When the page loads, it's not there. So I had just taken the screenshot after the page load.

Oh, I see. I think it's a bug that this text sticks around. It should disappear when you're not longer hovering over the bar that it refers to, right?

alya commented:
As another follow-up, can we drop the "Analytics documentation" link at the bottom of the page? That page doesn't have any information that's useful to you once you're already viewing the analytics.

alya commented:
Anyway, the two points above can be follow-ups -- looks good to me (in Chrome on a Mac)!

The other thing I noticed is that the charts don't look good on mobile (tested via dev tools). But this is not a regression, so I think we can just file an issue and figure it out later.

alya commented:
@timabbott Over to you.

timabbott commented:
Merged, thanks @laurynmm!
